src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbol.cs

```diff
@@ -1149,6 +1149,25 @@ private static void PartialMethodChecks(SourceOrdinaryMethodSymbol definition, S
         {
             Debug.Assert(!ReferenceEquals(definition, implementation));
 
+            MethodSymbol constructedDefinition = definition;
+            if (implementation.IsGenericMethod)
+            {
+                constructedDefinition = definition.Construct(implementation.TypeArgumentsWithAnnotations);
+            }
+
+            bool returnTypesEqual = constructedDefinition.ReturnTypeWithAnnotations.Equals(implementation.ReturnTypeWithAnnotations, TypeCompareKind.AllIgnoreOptions);
```

Comment:
Why did you choose this over `CLRSignatureCompareOptions`?

Reply:
Hmm the `TypeCompareKind` here appears to be different than how we verify parameters. For example tuple element names matter in parameter names. Unless there is a good reason otherwise think we should use the same comparisons for return type and parameters.

Reply:
I used `SourceMemberContainerSymbol.CheckOverrideMember` as a basis for this code:
https://github.com/dotnet/roslyn/blob/85b65a41b0069d5757039f5a5bdb8cb39e3af641/src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs#L939

Reply:
Think that makes sense in overrides because:

1. It's historically what we did so pretty much had to maintain it for consistency when adding new features. Basically it would be weird if we didn't error for `dynamic` vs. `object` but did error for `IntPtr` vs. `nint`. 
1. It's expected that the developer doesn't control the signature of the base member. Hence we should give them the flexbility to use `dynamic` in their code base because it makes sense while in the base case `object` was the more natural choice. Forcing them to use `object` in the override would just result in them adding a bunch of unnecessary casts.

For this feature neither is a concern. This is a new feature so compat doesn't come into play and the developer controls both definitions. 

Happy to hear if others disagree but my instinct is to say that parameters and return types should have the same comparisons done here.

Reply:
I may have misunderstood your comment, but it looks like CLRSignatureCompareOptions does not detect a difference between `object`/`dynamic` or `IntPtr`/`nint`. It looks like constructing a scenario which requires CLRSignatureCompareOptions involves custom modifiers, array sizes or lower bounds, which I am not sure how to do in source. Do you have any ideas for scenarios which illustrate the need for `CLRSignatureCompareOptions`?

> It's expected that the developer doesn't control the signature of the base member.

It feels like this is also the case to a significant degree when one of the 'partial' declarations comes from a code generator.

Reply:
>  I may have misunderstood your comment, but it looks like CLRSignatureCompareOptions does not detect a difference between object/dynamic or IntPtr/nint.

The suggestion of `CLRSignatureCmopareOptions` was just an example, didn't mean to suggest it should be used.

> It feels like this is also the case to a significant degree when one of the 'partial' declarations comes from a code generator.

True but it is also something that it's reasonable for the source generator to replicate. Consider that if they don't replicate what the user wrote, particularly the tuple element names, then they're subverting the will of the developer and making their generated code much harder to consume.

Reply:
> Consider that if they don't replicate what the user wrote, particularly the tuple element names, then they're subverting the will of the developer and making their generated code much harder to consume.

This applies when the source generator produces an implementing declaration: a well-behaved generator should just copy the signature of the user-authored defining declaration. But in the case where the generator produces a defining declaration and the user authors an implementing declaration, the user has to align with a generator that they likely do not control. Still, it's hard to picture a situation where a user is inconvenienced by being unable to substitute 'dynamic' for 'object' or 'IntPtr' for 'nint' or vice versa.

Probably the main reason to allow such differences for return types is probably that 'object'/'dynamic' etc. differences are currently allowed for parameters, so to only disallow them for returns is inconsistent.